rust/src/utils/kmeans.rs

```diff
@@ -417,20 +419,39 @@ impl KMeans {
         let dimension = self.dimension;
         let n = data.len() / self.dimension;
         let metric_type = self.metric_type;
-        let cluster_with_distances = stream::iter(0..n)
+        const CHUNK_SIZE: usize = 1024;
+        let cluster_with_distances = stream::iter((0..n).step_by(CHUNK_SIZE))
             // make tiles of input data to split between threads.
-            .chunks(1024)
             .zip(repeat_with(|| (data.clone(), self.centroids.clone())))
-            .map(|(indices, (data, centroids))| async move {
+            .map(|(start_idx, (data, centroids))| async move {
                 let data = tokio::task::spawn_blocking(move || {
-                    let dist = metric_type.batch_func();
+                    let array = data.values();
+                    let centroids_array = centroids.values();
                     let mut results = vec![];
-                    for idx in indices {
-                        let vector = data.slice(idx * dimension, dimension);
-                        let distances = dist(vector.values(), centroids.values(), dimension);
-                        let cluster_id = argmin(distances.as_ref()).unwrap();
-                        let distance = distances.value(cluster_id as usize);
-                        results.push((cluster_id, distance))
+                    for idx in start_idx..min(start_idx + CHUNK_SIZE, n) {
+                        let vector = &array[idx * dimension..(idx + 1) * dimension];
+                        let (cluster_id, distance) = {
+                            let mut min = std::f32::MAX;
+                            let mut min_idx = 0;
+                            for (idx, other) in centroids_array.chunks_exact(dimension).enumerate()
+                            {
+                                // We've found about 40% performance improvement by using static dispatch instead
+                                // of dynamic dispatch.
+                                //
+                                // NOTE: Please make sure run benchmark when changing the following code.
+                                // `RUSTFLAGS="-C target-cpu=native" cargo bench --bench ivf_pq`
+                                let dist = match metric_type {
+                                    MetricType::L2 => vector.l2(other),
+                                    MetricType::Cosine => vector.cosine(other),
+                                };
+                                if dist < min {
+                                    min = dist;
+                                    min_idx = idx;
+                                }
+                            }
+                            (min_idx, min)
+                        };
+                        results.push((cluster_id as u32, distance))
```

Comment:
`results` isn't pre-sized and `push` does bounds checking. So there's probably some performance improvement if we do 

```rust
(start_idx..min(start_idx + CHUNK_SIZE, n))
    .map(|idx| { ... })
    .collect_vec()
```

Reply:
Nice. Fixed.

We could probably push harder as we don't really need to store `(cluster_id, distance)` in vectors. We can just add them directly to the new centroids matrix.

Reply:
Yeah I imagine that will be quite difficult since you will need to split the centroids matrix into disjoint mutable slices that can be provided to each of the threads.

Reply:
```
CreateIVF256,PQ96(d=768)
                        time:   [29.153 s 29.237 s 29.319 s]
                        change: [-11.313% -10.839% -10.329%] (p = 0.00 < 0.10)
                        Performance has improved.
```

Happy report that with `collect::<Vec<_>>` it is another 10% speed up.

Reply:
> Yeah I imagine that will be quite difficult since you will need to split the centroids matrix into disjoint mutable slices that can be provided to each of the threads.

Another way would be, we first to run a `group_by` over the `cluster_ids` . And run the sum over each grouped list.